cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsCloudWatchService.java

```diff
@@ -71,20 +72,41 @@ public void addCloudWatchAlarmsForSystemFailures(List<CloudResource> instances,
     }
 
     public void deleteCloudWatchAlarmsForSystemFailures(CloudStack stack, String regionName, AwsCredentialView credentialView) {
-        List<CloudInstance> instances = stack.getGroups().stream()
-                .flatMap(group -> group.getInstances().stream()).collect(Collectors.toList());
+        List<String> instanceIds = stack.getGroups().stream()
+                .flatMap(group -> group.getInstances().stream())
+                .map(CloudInstance::getInstanceId)
+                .collect(Collectors.toList());
+        deleteCloudWatchAlarmsForSystemFailures(stack, regionName, credentialView, instanceIds);
+    }
+
+    public void deleteCloudWatchAlarmsForSystemFailures(CloudStack stack, String regionName, AwsCredentialView credentialView, List<String> instanceIds) {
         String isCreate = stack.getParameters().get(CLOUDWATCH_CREATE_PARAMETER);
         if (isCreate != null && isCreate.equals(Boolean.TRUE.toString())) {
-            instances.stream().forEach(instance -> {
-                try {
-                    DeleteAlarmsRequest deleteAlarmsRequest = new DeleteAlarmsRequest().withAlarmNames(instance.getInstanceId() + alarmSuffix);
-                    AmazonCloudWatchClient amazonCloudWatchClient = awsClient.createCloudWatchClient(credentialView, regionName);
-                    amazonCloudWatchClient.deleteAlarms(deleteAlarmsRequest);
-                    LOGGER.debug("Deleted cloudwatch alarm.-", instance.getInstanceId());
-                } catch (AmazonCloudWatchException acwe) {
-                    LOGGER.error("Unable to delete cloudwatch alarm for instanceId {}: {}", instance.getInstanceId(), acwe.getLocalizedMessage());
-                }
-            });
+            List<String> instanceIdsFromStack = stack.getGroups().stream()
+                    .flatMap(group -> group.getInstances().stream())
+                    .map(CloudInstance::getInstanceId)
+                    .collect(Collectors.toList());
+            List<String> instanceIdsNotInStack = instanceIds.stream()
+                    .filter(instanceId -> !instanceIdsFromStack.contains(instanceId))
+                    .collect(Collectors.toList());
+            if (!instanceIdsNotInStack.isEmpty()) {
+                LOGGER.error("Instance IDs [{}] are not part of stack {}", instanceIdsFromStack, stack);
+                throw new CloudConnectorException("Unable to delete cloud watch alarms for system failure because instances ID are not a part of the stack.");
+            }
+            deleteCloudWatchAlarmsForSystemFailures(regionName, credentialView, instanceIds);
         }
     }
+
+    private void deleteCloudWatchAlarmsForSystemFailures(String regionName, AwsCredentialView credentialView, List<String> instanceIds) {
+        instanceIds.stream().forEach(instanceId -> {
+            try {
+                DeleteAlarmsRequest deleteAlarmsRequest = new DeleteAlarmsRequest().withAlarmNames(instanceId + alarmSuffix);
```

Comment:
Although it is possible to request up to 100 alarms to be deleted, "In the event of an error, no alarms are deleted." - https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_DeleteAlarms.html

We could do something where we partition the list into sets of 100 and try to delete them (but this is not needed in the case of FreeIPA). If there is a failure then we could fall back to this implementation, that way we ensure all alarms that can be deleted are deleted. If the alarm deletion was used by datahub, then this should be done. But its currently only enabled for FreeIPA. With FreeIPA management service, there should be at most 3 instances in a delete request since there is a 4 node max. But in most cases, I suspect we will rarely get to cases where 4 nodes are used and 3 of them have failed. I am not sure the complexity is worth it here.

Reply:
so for FreeIPA this is not an issue, but my fear is if we might introduce this for clusters for some reason, we won't check if it scales well. So my suggestion here is to make bulk delete, and partition it by a 100.
Also the current way is not too idempotent as if we have to retry this state and like half of the alarm delete it will fail when we try to delete a non existing one.
So we should use describeAlarms or similar to fetch which exists and try to delete them. Also we might want to fail deletion if we can't delete them, because we will leave trash behind.

Reply:
I created CB-7340 for this.